CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,475 @@
+---
+cip: <to be assigned>
+title: Extension to ODIS for Password Hashing
+author: Victor Graf <victor@clabs.co>, Cody Born <cody@clabs.co>, Alec Schaefer <alec@clabs.co>
+discussions-to: https://github.com/celo-org/celo-proposals/issues/233
+status: Draft
+type: Standards Track
+category: 3
+created: 2021-06-15
+<!--
+  TODO(victor): Once a CIP for the cryptography is ready, add that here as a dependency.
+  requires:
+-->
+license: Apache 2.0
+---
+
+### Terminology
+
+<!-- TODO(victor): Actually go through and add these keywords where appropriate -->
+The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to be interpreted as
+described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html).
+
+## Simple Summary
+
+An extension to the ODIS phone number privacy service to support granular rate limiting with the
+goal of hardening passwords for use as encryption and authorization keys.
+
+## Abstract
+
+<!-- TODO(victor): Once we have a roughly agreed upon proposal for the pOPRF, link it here -->
+Given the required cryptographic functionality (to be discussed in more detail in a future
+proposal), this proposal describes an interface to be exposed by the ODIS service allowing the user
+to specify a domain string (i.e. a salt value) with rate-limiting rules to be combined with their
+blinded input (e.g.  phone number or password).
+
+Central to the proposed interface are domains. A domain is a struct which specifies rate limiting
+information and uniquely identifies the context of the secret to be hashed. As an example, a domain
+for hashing an account password might specify the username (context) and a cap of 10 password
+attempts (rate-limiting parameter). These would be combined with their password to form the final
```

Comment:
Is the idea to have the user set a custom rate limit for each domain? Why would this be preferable to having a predetermined constant rate limit (perhaps set by Governance) for each domain type (e.g. password, phone number)? It seems like this could introduce unwanted complexity on the user's end, and could make the rate-limiting scheme ineffective if the user doesn't know what they are doing (e.g. they set a super high number of max password attempts)?

Reply:
Ah, seems like from the `Domain Specification` section below, that users can't specify custom rate limits and rather the ODIS determines the rate limit based on the Celo-wide rule set for the specified domain?

Reply:
When I say user in this doc, I generally mean the user of ODIS / the application developer. I would expect each application to choose a rate limit appropriate their needs. It's better than setting it centrally because each application may have different security requirements and we don't want to involve Governance where there is not a collective action problem. (i.e. it's perfectly acceptable for different applications / users to use different rate limits and domains, so there is no reason to force them all to be chosen centrally)

Reply:
To clarify on your question, the supported rule sets would be developed and supported centrally as part of the ODIS software, but the parameters for those rulesets can be controlled by the ODIS user. Of course, this blurs a little when your consider some interesting use cases, such as the smart contract domain example, where the parameter is essentially the ruleset.